algorithms/binary-search.py

```diff
@@ -0,0 +1,88 @@
+def bin_search_recursive(arr, val, left=0, right=None):
+  '''
+  Very verbose and error-prone because of many ifs solution.
+  '''
+  if right == None:
+    right = max(len(arr) - 1, 0)
+
+  if right - left == 1:
+    if arr[left] == val:
+      return left
+    elif arr[right] == val:
+      return right
+    else:
+      return -1
+  elif right - left == 0:
+    return -1
+
+  middle = (left + right) // 2
+  if val <= arr[middle]:
+    return bin_search_recursive(arr, val, left, middle)
+  else:
+    return bin_search_recursive(arr, val, middle, right)
+
+
+def bin_search(arr, val):
```

Comment:
This is slightly better but still base case is too long: compare to left and to right, else return -1.